pkg/handlers/middleware.go

```diff
@@ -44,3 +51,96 @@ func RequireValidLicenseIDMiddleware(next http.Handler) http.Handler {
 		next.ServeHTTP(w, r)
 	})
 }
+
+// Code for the cache middleware
+type CacheEntry struct {
+	RequestBody  []byte
+	ResponseBody []byte
+	StatusCode   int
+	Expiry       time.Time
+}
+
+type cache struct {
+	store map[string]CacheEntry
+	mu    sync.RWMutex
+}
+
+func NewCache() *cache {
+	return &cache{
+		store: map[string]CacheEntry{},
+	}
+}
+
+func (c *cache) Get(key string) (CacheEntry, bool) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	entry, found := c.store[key]
+	if !found || time.Now().After(entry.Expiry) {
+		return CacheEntry{}, false
+	}
+	return entry, true
+}
+
+func (c *cache) Set(key string, entry CacheEntry, duration time.Duration) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	entry.Expiry = time.Now().Add(duration)
+	c.store[key] = entry
+}
+
+type responseRecorder struct {
+	http.ResponseWriter
+	Body       *bytes.Buffer
+	StatusCode int
+}
+
+func (r *responseRecorder) WriteHeader(code int) {
+	r.StatusCode = code
+	r.ResponseWriter.WriteHeader(code)
+}
+
+func (r *responseRecorder) Write(b []byte) (int, error) {
+	r.Body.Write(b)
+	return r.ResponseWriter.Write(b)
+}
+
+func CacheMiddleware(next http.HandlerFunc, duration time.Duration) http.HandlerFunc {
+	// Each handler has its own cache to reduce contention for the in-memory store
+	cache := NewCache()
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		body, err := io.ReadAll(r.Body)
+		if err != nil {
+			logger.Error(errors.Wrap(err, "cache middleware failed ready request body"))
+			http.Error(w, "cache middleware: unable to read request body", http.StatusInternalServerError)
+			return
+		}
+
+		r.Body = io.NopCloser(bytes.NewBuffer(body))
+
+		key := r.Method + "::" + r.URL.Path + "::" + string(body)
```

Comment:
how about using a hash of the concatenated bytes to avoid encoding issues?

Reply:
Good call.  I'm curious how it would lead to encoding issues?  Aren't all strings in Go UTF-8 encoded?  Additionally, the key is never parsed out once concatenated.

Reply:
+1 for using digests.  Pods get kicked out when memory limit is reached, and payloads can be large.

Reply:
Will SHA256 the key!

Reply:
It's a bit awkward here, but if payload is json, I would suggest remarshaling it to ensure consistent key ordering, indentation, and line breaks.

Reply:
JSON ordering issues would throw off the cache. Good cache.

Reply:
@divolgin Here is the diff for payload comparison: https://github.com/replicatedhq/replicated-sdk/pull/193/commits/ce888109bda361f9a830eaff209fe25f01988051